src/main/java/org/commonjava/util/partyline/FileTree.java

```diff
@@ -327,46 +327,55 @@ boolean tryLock( File file, String ownerName, String label, LockLevel lockLevel,
         while ( end < 1 || System.currentTimeMillis() < end )
         {
             T result = tryLock( realFile, Thread.currentThread().getName(), "Open File for " + ( doOutput ? "output" : "input" ),
-                                           doOutput ? LockLevel.write : read, WAIT_TIMEOUT, unit, ( opLock ) -> {
+                                           doOutput ? LockLevel.write : read, timeout, unit, ( opLock ) -> {
                         FileEntry entry = entryMap.get( realFile.getAbsolutePath() );
-                        if ( entry.file != null )
+                        boolean proceed = false;
+                        synchronized ( entry )
```

Comment:
One thing that might work is to modify withOpLock() to use a while loop for the full timeout, then use a very small timeout in the lock() call inside...and put the synchronized block where we pull out the ReentrantLock inside the while loop. Then, we could let JoinableFile.close() remove the ReentrantLock from the map, and change the map to something like ConcurrentHashMap.

If the while loop has some sort of wait step after calling (and failing on) lock(), then on the next pass it would detect the removal of ReentrantLock and create a new one.

That's a bit heavy, but it might work. @ruhan1 does that make sense?